agent/consul/server.go

```diff
@@ -889,6 +890,34 @@ func NewServer(config *Config, flat Deps, externalGRPCServer *grpc.Server,
 	// to enable RPC forwarding.
 	s.grpcLeaderForwarder = flat.LeaderForwarder
 
+	// Start watching HCP Link resource. This creates a channel that we can use to
+	// start and stop HCP manager when appropriate. This needs to be created after
+	// the GRPC services are set up in order for the resource service client to
+	// function. This uses the insecure grpc channel so that it doesn't need to
+	// present a valid ACL token.
+	//
+	// If this fails, HCP linking will not work, but to avoid crashing Consul, we log
+	// the error and continue on.
+	hcpLinkWatchCh, err := hcpctl.NewLinkWatch(
+		&lib.StopChannelContext{StopCh: shutdownCh},
+		logger.Named("hcp-link-watcher"),
+		pbresource.NewResourceServiceClient(s.insecureSafeGRPCChan),
+	)
+	if err != nil {
+		s.logger.Error("HCP Link watcher failed to start. HCP Link functionality is disabled", "error", err)
+	} else {
+		go hcp.MonitorHCPLink(
+			&lib.StopChannelContext{StopCh: shutdownCh},
+			logger.Named("hcp-link-watcher"),
+			s.hcpManager,
+			hcpLinkWatchCh,
+			hcpclient.NewClient,
+			bootstrap.LoadManagementToken,
+			flat.HCP.Config,
+			flat.HCP.DataDir,
+		)
+	}
```

Comment:
The separation between the "link watcher" and the "link monitor" seems like it could be vulnerable to first time failure. 

IIUC, if L901 fails for any reason, the one-shot attempt to successfully link the cluster has been used up won't be retried unless the server is restarted.

I understand this makes it easier to test, but consider collapsing the two go routines into a simpler single goroutine that includes a perpetual retry with backoff of `WatchList` to handle errors and as a result doesn't require `hcpLinkWatchCh`.

Reply:
Ok, that's a good point. I can take a stab at that!

Reply:
@analogue I ended up keeping 2 separate functions to keep the ease of testing, but combined them into 1 top-level function that can be called from a go-routine in server.go. Let me know what you think!

Reply:
From what I can tell, you still have two go routines, and `WatchList` won't ever recover if it fails after a previous success during the lifetime of the server.

Reply:
I do still have 2 goroutines, but I have this retry loop which guarantees the `WatchClient` from `WatchList` gets created successfully if it finishes:
```golang
	var watchClient pbresource.ResourceService_WatchListClient
	watchListErrorBackoff := &retry.Waiter{
		MinFailures: 1,
		MinWait:     1 * time.Second,
		MaxWait:     1 * time.Minute,
	}
	for {
		var err error
		watchClient, err = client.WatchList(
			ctx, &pbresource.WatchListRequest{
				Type:       pbhcp.LinkType,
				NamePrefix: hcpctl.LinkName,
			},
		)
		if err != nil {
			logger.Error("failed to create watch on Link", "error", err)
			watchListErrorBackoff.Wait(ctx)
			continue
		}

		break
	}
```

Are you saying there's some error state that can happen after this is successfully created that would require us to create a new `WatchClient`?

Reply:
Correct. When a stream returns an error, you need to re-establish the stream via `WatchList`. 

https://github.com/grpc/grpc-go/blob/master/stream.go#L136

Reply:
Ah, okay I see. I'll try that out. If it's taking too long and you consider this non-blocking, I may shelf this for now.